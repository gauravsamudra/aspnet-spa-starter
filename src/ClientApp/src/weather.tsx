import React, { useState, useEffect } from "react";

export interface IProps {
  id: number;
}

export interface IWeather {
  summary: string;
  temperatureC: number;
  date: string;
}

export function Weather(props: IProps) {
  const [weather, setWeather] = useState(null);

  async function fetchWeather(id) {
    const response = await fetch(`/WeatherForecast/${id}/`);
    setWeather(await response.json());
  }
  useEffect(() => {
    fetchWeather(props.id);
  }, [props.id]);

  if (!weather) {
    return <div>loading...</div>;
  }

  return (
    <div>
      <h2 className="summary">Forecast: {weather.summary}</h2>
      <strong>{weather.temperatureC}</strong> Â°C
      <br />
      Date: {weather.date}
    </div>
  );
}