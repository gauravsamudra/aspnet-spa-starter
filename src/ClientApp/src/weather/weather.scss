.Weather { 
    table {
        border-spacing: 0px;
        border-style: solid;
        border-width: 1px;
        border-color: #e0e0dc;
        
        td {
            padding-left: 18px;
            padding-right: 18px;
            padding-top: 6px;
            padding-bottom: 6px;
            border: 1px solid #e0e0dc;
        }
        th {
            padding-left: 18px;
            padding-right: 18px;
            padding-top: 6px;
            padding-bottom: 6px;
            border: 1px solid #e0e0dc;
            background-color: rgba(212,221,228,.5);
        }
        caption {
            padding-top: 6px;
            caption-side: bottom;
        }
    }    
}