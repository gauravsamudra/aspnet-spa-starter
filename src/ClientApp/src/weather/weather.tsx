import React, { useState, useEffect } from "react";
import './weather.scss'
export interface IProps {
}

export interface IWeather {
  summary: string;
  temperatureC: number;
  temperatureF: number;
  date: string;
}

export function Weather(props: IProps) {
  const [weatherForecasts, setWeatherForecasts] = useState([] as IWeather[]);

  async function fetchWeatherForecasts() {
    const response = await fetch('/WeatherForecast/');
    if (!response.ok) {
      return;
    }
    let weatherForecasts = await response.json();
    weatherForecasts.map(w => w.date = new Date(w.date).toDateString());
    setWeatherForecasts(weatherForecasts);
  }
  useEffect(() => {
    fetchWeatherForecasts();
  }, [props]);

  return (
    <div className="Weather">
      <table>
        <thead>
          <tr>
            <th>Date</th>
            <th>Temp. °C</th>
            <th>Temp. °F</th>
            <th>Summary</th>
          </tr>
        </thead>
        <tbody>
          {weatherForecasts.map((forecast, index) =>
            <tr key={index}>
              <td>{forecast.date}</td>
              <td>{forecast.temperatureC}</td>
              <td>{forecast.temperatureF}</td>
              <td>{forecast.summary}</td>
            </tr>
          )}
        </tbody>
        <caption>Weather forecast</caption>
      </table>
    </div>
  );
}