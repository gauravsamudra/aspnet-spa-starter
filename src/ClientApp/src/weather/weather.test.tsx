import { render } from '@testing-library/react';
import React from "react";
import { Weather, IWeather } from "./weather";

describe('Weather forecast', () => {
  it('should look sensible', () => {
    const { container } = render(<Weather />);
    expect(container).toMatchSnapshot();
  });
  
  it('should fetch weather from API', () => {
  
    const expectedValue = [{
      date: new Date('2019-04-05').toDateString(),
      temperatureC: -5,
      temperatureF: -30,
      summary: 'Test London mild weather'
    } as IWeather];

    const fetchSpy = jest.spyOn(window, 'fetch').mockImplementation((input) => {
      if (input === '/WeatherForecast/') {
        return Promise.resolve({ 
          json: () => Promise.resolve(expectedValue)
        } as Response);
      }
      else
        return Promise.reject();
    });
    try {
      render(<Weather />);

      expect(fetchSpy).toHaveBeenCalledTimes(1);
    }
    finally {
      fetchSpy.mockReset();
    }
  });
})
