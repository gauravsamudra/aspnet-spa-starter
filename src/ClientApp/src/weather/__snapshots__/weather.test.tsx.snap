// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Weather forecast should look sensible 1`] = `
<div>
  <div
    class="Weather"
  >
    <table>
      <thead>
        <tr>
          <th>
            Date
          </th>
          <th>
            Temp. °C
          </th>
          <th>
            Temp. °F
          </th>
          <th>
            Summary
          </th>
        </tr>
      </thead>
      <tbody />
      <caption>
        Weather forecast
      </caption>
    </table>
  </div>
</div>
`;
